src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py

```diff
@@ -0,0 +1,432 @@
+        else:
+            _find_layers_by_name(
+                module, self._student_layer_names, cached_student_layers
+            )
+        _LOGGER.info("Distilling student layers: %s", self._student_layer_names)
+
+        cached_teacher_layers: Dict[str, torch.nn.Module] = {}
+        if self._teacher_layer_names is None:
+            _find_layers_by_type(self._teacher, cached_teacher_layers)
+            self._teacher_layer_names = list(cached_teacher_layers.keys())
+        else:
+            _find_layers_by_name(
+                self._teacher, self._teacher_layer_names, cached_teacher_layers
+            )
+        _LOGGER.info("Distilling teacher layers: %s", self._teacher_layer_names)
+
+        if len(self._teacher_layer_names) != len(self._student_layer_names):
+            raise ValueError(
+                "Found different numbers of teacher and student layers to distill. "
+                "Set teacher_layer_names and student_layer_names explicitly."
+            )
+
+        self._student_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_student_output, self._student_output_shapes
+                )
+            )
+            for name, layer in cached_student_layers.items()
+        ]
+
+        self._teacher_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_teacher_output, self._teacher_output_shapes
+                )
+            )
+            for name, layer in cached_teacher_layers.items()
+        ]
+
+    def update(
+        self,
+        module: Module,
+        optimizer: torch.optim.Optimizer,
+        epoch: float,
+        steps_per_epoch: int,
+    ):
+        super().update(module, optimizer, epoch, steps_per_epoch)
+        if epoch >= self.end_epoch:
+            for handle in self._student_handles:
+                handle.remove()
+            for handle in self._teacher_handles:
+                handle.remove()
+            self._cached_student_output.clear()
+            self._cached_teacher_output.clear()
+            self._student_output_shapes.clear()
+            self._teacher_output_shapes.clear()
+            self._student_handles.clear()
+            self._teacher_handles.clear()
+            self._projection.clear()
+
+    def compute_total_loss(self, loss, distillation_loss):
+        return loss + self.gain * distillation_loss
+
+    def compute_distillation_loss(self, optimizer: torch.optim.Optimizer, **kwargs):
+        if self.project_features and len(self._projection) == 0:
+            # NOTE: have to call initialize here because we need the cached output
+            # from the module. i.e. we need forward to have been called already
+            projection_params = self._initialize_projection()
+            optimizer.add_param_group({"params": projection_params})
```

Comment:
noting that params is expected by pytorch

Reply:
what do you mean by this? is this a reserved name? why not name the group `distill_projection_params`?

Reply:
@bfineran params is the keyword that pytorch requires you to use (all param groups is a dict of keys like "params", "lr", etc)